bin/oneoffs/migrate_storage.py

```diff
@@ -0,0 +1,216 @@
+#!/usr/bin/env python
+import argparse
+import datetime
+import logging
+import os
+import uuid
+
+import fs.path
+import fs.move
+
+from api import config, util
+
+log = logging.getLogger('migrate_storage')
+log.setLevel(logging.INFO)
+
+COLLECTIONS_PREFIXES = [('projects', 'files'),
+                            ('acquisitions', 'files'),
+                            ('analyses', 'files'),
+                            ('sessions', 'files'),
+                            ('sessions', 'subject.files'),
+                            ('collections', 'files')]
+
+
+class MigrationError(Exception):
+    pass
+
+
+def get_files_by_prefix(document, prefix):
+    for key in prefix.split('.'):
+        document = document.get(key, {})
+    return document
+
+
+def show_progress(current_index, total_files):
+    if current_index % (total_files / 10 + 1) == 0:
+        log.info('Processed %s files of total %s files ...' % (current_index, total_files))
+
+
+def cleanup_empty_folders():
+    log.info('Cleanup empty folders')
+
+    for _dirpath, _, _ in os.walk(config.get_item('persistent', 'data_path'), topdown=False):
+        if not (os.listdir(_dirpath) or config.get_item('persistent', 'data_path') == _dirpath):
+            os.rmdir(_dirpath)
+
+
+def get_collections_files():
+    _files = []
+
+    for collection, prefix in COLLECTIONS_PREFIXES:
+        cursor = config.db.get_collection(collection).find({})
+        for document in cursor:
+            for f in get_files_by_prefix(document, prefix):
+                f_dict = {
+                    'collection_id': document.get('_id'),
+                    'collection': collection,
+                    'fileinfo': f,
+                    'prefix': prefix
+                }
+                _files.append(f_dict)
+
+    return _files
+
+
+def get_gears_files():
+    cursor = config.db.get_collection('gears').find({})
+    _files = []
+
+    for document in cursor:
+        if document.get('exchange', {}).get('git-commit', '') == 'local':
+            f_dict = {
+                'gear_id': document['_id'],
+                'gear_name': document['gear']['name'],
+                'exchange': document['exchange']
+            }
+            _files.append(f_dict)
+
+    return _files
+
+
+def migrate_collections():
+    log.info('Migrate collection files...')
+
+    _files = get_collections_files()
+
+    for i, f in enumerate(_files):
+        try:
+            file_id = f['fileinfo'].get('_id', '')
+            if file_id:
+                file_path = util.path_from_uuid(file_id)
+                if not config.fs.isfile(file_path):
```

Comment:
I assume here is where files that have been created since starting the conversion would be ignored? Also does this mean the conversion script can run many times and will just be a noop when no files need to be migrated?

Reply:
Yes, exactly